tfhe/src/integer/server_key/radix/scalar_add.rs

```diff
@@ -50,25 +55,20 @@ impl ServerKey {
     /// ciphertext.
     ///
     /// The result is assigned to the `ct_left` ciphertext.
-    pub fn unchecked_scalar_add_assign<PBSOrder: PBSOrderMarker>(
+    pub fn unchecked_scalar_add_assign<PBSOrder, T>(
         &self,
         ct: &mut RadixCiphertext<PBSOrder>,
-        scalar: u64,
-    ) {
-        // Bits of message put to 1
-        let mask = (self.key.message_modulus.0 - 1) as u64;
-
-        let mut power = 1_u64;
-        // Put each decomposition into a new ciphertext
-        for ct_i in ct.blocks.iter_mut() {
-            let mut decomp = scalar & (mask * power);
-            decomp /= power;
-
-            self.key.unchecked_scalar_add_assign(ct_i, decomp as u8);
-
-            //modulus to the power i
-            let Some(new_power) = power.checked_mul(self.key.message_modulus.0 as u64) else {break};
-            power = new_power;
+        scalar: T,
+    ) where
+        PBSOrder: PBSOrderMarker,
+        T: DecomposableInto<u8>,
+    {
+        let bits_in_message = self.key.message_modulus.0.ilog2();
+        let decomposer =
+            BlockDecomposer::with_early_stop_at_zero(scalar, bits_in_message).iter_as::<u8>();
+        for (ciphertext_block, scalar_block) in ct.blocks.iter_mut().zip(decomposer) {
+            self.key
+                .unchecked_scalar_add_assign(ciphertext_block, scalar_block);
         }
```

Comment:
Yes but I prefer to still write algorithm that rely on the fact that all blocks share the same message modulus,
then once we really need to add support for multiple message mod blocks we will work on all algorithms.

Just like to add signed integers support we will work again on all algorithms.

I think its going to be easier to manage: as of now no signed integer support at all, and no support for mixed message modulus

Reply:
Its an OKR on its own

Reply:
Fine by me

Reply:
Just means a few algorithms are broken for parameters we provide

Reply:
Hmm, I would not say that
While true, it is not an intended use case of the current API of the crate

- As we don't provide functions that allow to encrypt while mixing message moduluses at the integer level, that means to have an integer ciphertext with mixed message modulus, the user would have to fiddle with the internals of the integer ciphertext after encryption or create its own encryption function using shortint API, construct a RadixCiphertext from these shortint, so then expecting the integer level to work is something else.
And of course by fidling the internals of the ciphertext I can break almost all functions available in the api.

- There is no mention of that possibility in the doc

- And more importantly, in my opinion, mixed message modulus was always broken since there are no tests for it, and with how easy it is to miss a corner case or simply not write a good algorithm or break existing stuff while trying to improve it, no tests == does not work == not supported

Reply:
All I’m saying is anybody using 3_3 parameters with integers may encounter issues while not doing anything fancy :) and yes it won’t be fixed overnight and will require a big chunk of work

Reply:
I think only comparisons are known to be truly broken on 3_3, other op have tests which passes, maybe i'm more optimist on this than you 😁

Reply:
Just a glass half empty kind of guy 😄

Reply:
To be fair, 3_3 is fancy, with the way we encrypt by using 3_3 you can't represent any of the common integer sizes (u8, u16, etc) you always end up with more bits, so a different modulus than if you compare it the the closes `uX` type, but our algorithms still works with 3_3, you just can't do arithmetic on (3 * 43 = 129 bits) and expects results to be the same as arithemtic over u128 its just not the same thing